edk2basetools/VfrCompiler/IfrUtility.py

```diff
@@ -0,0 +1,2238 @@
+        self.NewDataType.TotalSize = Size
+
+    def SetNewTypeAlign(self, Align):
+        self.NewDataType.Align = Align
+
+    def SetNewTypeName(self, TypeName):
+        if self.NewDataType == None:
+            return VfrReturnCode.VFR_RETURN_ERROR_SKIPED
+
+        if TypeName == None:
+            return VfrReturnCode.VFR_RETURN_FATAL_ERROR
+
+        if len(TypeName) >= MAX_NAME_LEN:
+            return VfrReturnCode.VFR_RETURN_INVALID_PARAMETER
+
+        pType = self.DataTypeList
+        while pType != None:
+            if pType.TypeName == TypeName:
+                return VfrReturnCode.VFR_RETURN_REDEFINED
+            pType = pType.Next
+
+        self.NewDataType.TypeName = TypeName
+        return VfrReturnCode.VFR_RETURN_SUCCESS
+
+    def AlignStuff(self, Size, Align):
+        return Align - (Size) % (Align)
+
+    def DeclareDataTypeEnd(self):
+        if self.NewDataType.TypeName == "":
+            return
+
+        if self.NewDataType.TotalSize % self.NewDataType.Align != 0:
+            self.NewDataType.TotalSize += self.AlignStuff(self.NewDataType.TotalSize, self.NewDataType.Align)
+
+        self.RegisterNewType(self.NewDataType)
+        if self.FirstNewDataTypeName == None:
+            self.FirstNewDataTypeName = self.NewDataType.TypeName
+
+        self.NewDataType = None
+
+    # two definitions
+    def GetDataTypeSizeByTypeName(self, TypeName):
+        Size = 0
+        pDataType = self.DataTypeList
+        while pDataType != None:
+            if pDataType.TypeName == TypeName:
+                Size = pDataType.TotalSize
+                return Size, VfrReturnCode.VFR_RETURN_SUCCESS
+            pDataType = pDataType.Next
+
+        return Size, VfrReturnCode.VFR_RETURN_UNDEFINED
+
+    def GetDataTypeSizeByDataType(self, DataType):
+        Size = 0
+        DataType = DataType & 0x0F
+        # For user defined data type, the size can't be got by this function.
+        if DataType == EFI_IFR_TYPE_OTHER:
+            return Size, VfrReturnCode.VFR_RETURN_SUCCESS
+        pDataType = self.DataTypeList
+        while pDataType != None:
+            if DataType == pDataType.Type:
+                Size = pDataType.TotalSize
+                return Size, VfrReturnCode.VFR_RETURN_SUCCESS
+            pDataType = pDataType.Next
+
+        return Size, VfrReturnCode.VFR_RETURN_UNDEFINED
+
+    def ExtractStructTypeName(self, VarStr):
+        try:
+            index = VarStr.index(".")
+        except ValueError:
+            return VarStr, len(VarStr)
+        else:
+            return VarStr[0:index], index + 1
+
+    def ExtractFieldNameAndArrary(self, VarStr, s):
+        ArrayIdx = INVALID_ARRAY_INDEX
+        s_copy = s
+        while s < len(VarStr) and VarStr[s] != "." and VarStr[s] != "[" and VarStr[s] != "]":
+            s += 1
+
+        FName = VarStr[s_copy:s]
+
+        if s == len(VarStr):
+            return ArrayIdx, s, FName, VfrReturnCode.VFR_RETURN_SUCCESS
+
+        elif VarStr[s] == ".":
```

Comment:
All of these can be individual if statements because you return if the condition is met.